### PR TITLE
Improve performance of LDAP matching

### PR DESCRIPTION
**Pull request by achristoforides**
This PR improves the performance of LDAP matching.

The performance improvements in https://github.com/CppMicroServices/CppMicroServices/pull/699 are characteristic of the performance gains exhibited by this PR except the performance gains in this PR are in-fact faster (by ~80ns on win64 when benchmarked on my machine, and generally faster on the other platforms).

Signed-off-by: The MathWorks, Inc. <alchrist@mathworks.com>

### CODE REVIEW

File: framework/src/util/Properties.cpp [hunk 1]
```diff
@@ -24,99 +24,140 @@
...
+Properties::Properties(const AnyMap& p)
+  : props(p)
 {
-  keys = std::move(o.keys);
-  values = std::move(o.values);
-  return *this;
+  if (p.GetType() != AnyMap::UNORDERED_MAP_CASEINSENSITIVE_KEYS) {
+    props_check::ValidateAnyMap(p);
+
+    PopulateCaseInsensitiveLookupMap();
```
Comment (achristoforides) [inline framework/src/util/Properties.cpp:62]: Do the validation check as part of PopulateCaseInsensitiveLookupMap() (if faster...)

File: framework/src/util/PropsCheck.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ * those stored in Property objects have.
+ * 
+ * The provided AnyMap is said to be valid if there exists no pairs of two keys
+ * which differ in case only (e.g., "service.feature", "Service.feature"). If this
+ * condition is not true, this function throws as defined below.
+ * 
+ * @param am The AnyMap to validate
+ * @throws std::runtime_error Thrown when `am` is invalid (described above)
+ */
+void ValidateAnyMap(const cppmicroservices::AnyMap& am)
```
Comment (achristoforides) [inline framework/src/util/PropsCheck.cpp:42]: Check to see if using a std::unordered_set (case-insensitive) is a better solution for checking validity.

File: framework/src/util/Properties.cpp [hunk 3]
```diff
@@ -24,99 +24,140 @@
...
+
+    if (!matchCase) {
+      auto ciItr = caseInsensitiveLookup.find(key);
+      if (ciItr != caseInsensitiveLookup.end()) {
+        return std::make_pair(
+          props.findUO_TypeChecked(ciItr->second.data())->second, true);
+      } else {
+        return std::make_pair(emptyAny, false);
+      }
     }
```
Comment (achristoforides) [inline framework/src/util/Properties.cpp:123]: This needs to happen in the ORDERED_MAP case.

File: framework/src/util/Properties.cpp [hunk 4]
```diff
@@ -24,99 +24,140 @@
...
+          props.findUOCI_TypeChecked(ciItr->second.data())->second, true);
+      } else {
+        return std::make_pair(emptyAny, false);
+      }
+    }
+
+    return std::make_pair(emptyAny, false);
+  } else {
+    throw std::runtime_error("Unknown AnyMap type.");
   }
```
Comment (achristoforides) [inline framework/src/util/Properties.cpp:145]: This logic should be for ORDERED_MAP

File: framework/include/cppmicroservices/AnyMap.h [hunk 5]
```diff
@@ -221,8 +223,11 @@ class US_Framework_EXPORT any_map
 
   any_map(map_type type);
```
Comment (pelliott-mathworks) [inline framework/include/cppmicroservices/AnyMap.h:224]: Some additional comments to explain some of the background and reasoning for the decisions that were made to improve efficiency would be helpful.

File: framework/src/util/PropsCheck.cpp (hunk 2 above)
Comment (achristoforides) [reply framework/src/util/PropsCheck.cpp:42]: It's not.

**Review by jeffdiclemente — APPROVED** [review]

**Pull request merged**